source/opt/eliminate_dead_io_components_pass.cpp

```diff
@@ -236,7 +236,27 @@ void EliminateDeadIOComponentsPass::ChangeIOVarStructLength(Instruction& io_var,
     }
     type_mgr->AttachDecoration(*dec, &new_struct_ty);
   }
-  analysis::Type* reg_new_var_ty = type_mgr->GetRegisteredType(&new_struct_ty);
+  // Clone name instructions for new struct type
+  std::vector<std::unique_ptr<Instruction>> names_to_add;
+  analysis::Type* reg_new_str_ty = type_mgr->GetRegisteredType(&new_struct_ty);
+  uint32_t new_struct_ty_id = type_mgr->GetTypeInstruction(reg_new_str_ty);
+  auto names = context()->GetNames(old_struct_ty_id);
+  for (auto n : names) {
+    Instruction* old_name_inst = n.second;
+    if (old_name_inst->opcode() == spv::Op::OpMemberName) {
+      auto midx = old_name_inst->GetSingleWordInOperand(1);
+      if (midx >= length) continue;
+    }
+    std::unique_ptr<Instruction> new_name_inst(old_name_inst->Clone(context()));
+    new_name_inst->SetInOperand(0, {new_struct_ty_id});
+    names_to_add.push_back(std::move(new_name_inst));
+  }
+  // We can't add the new names when we are iterating over name ranges above.
```

Comment:
Why that? Is it because the instruction we point to could be deleted? Or the iterator invalidated? (GetNames() comment or AddDebug2Inst() doesn't say it invalidates iterators)

Reply:
names is an iterator range pointing into a multimap. AddDebug2Inst adds to that multimap, which invalidates any iterators into that multimap. So iterating across the multimap has to be complete before adding to it.